clients/src/main/java/org/apache/kafka/common/utils/ByteUtils.java

```diff
@@ -392,12 +392,33 @@ public static void writeDouble(double value, ByteBuffer buffer) {
      * @param value The signed value
      */
     public static int sizeOfUnsignedVarint(int value) {
-        int bytes = 1;
-        while ((value & 0xffffff80) != 0L) {
-            bytes += 1;
-            value >>>= 7;
-        }
-        return bytes;
+        int leadingZeros = Integer.numberOfLeadingZeros(value);
+
+        // magic sequence of numbers that produces a function equivalent to this lookup
+        // table, where the index in the lookup table is provided by the number of
+        // leading zeros of the value, and the result is the number of bytes used
+        // in the output
+
+        // see the test cases as well to verify the implementation matches the prior
+        // for-loop logic
+
+        // final static byte[] LEADING_ZEROS_TO_U_VARINT_SIZE = new byte[] {
+        //     // 32 bits, and each 7-bits adds one byte to the output
+        //     5, 5, 5, 5, // 32
+        //     4, 4, 4, 4, 4, 4, 4, // 28
+        //     3, 3, 3, 3, 3, 3, 3, // 21
+        //     2, 2, 2, 2, 2, 2, 2, // 14
+        //     1, 1, 1, 1, 1, 1, 1, // 7
+        //     1 // 0
+        // };
+
+        // this is the core logic, but the Java encoding is suboptimal when we have a narrow
+        // range of integers, so we can do better here
+
+        // return (38 - leadingZeros) / 7 + leadingZeros / 32;
+
+        int leadingZerosBelow38DividedBy7 = ((38 - leadingZeros) * 0b10010010010010011) >>> 19;
+        return leadingZerosBelow38DividedBy7 + (leadingZeros >>> 5);
```

Comment:
Very nice. Regarding the comment, do we need the lookup table? Or can we explain it based on the core logic comment?